proposals/0000-th-stage-hygiene.rst

```diff
@@ -0,0 +1,680 @@
+
+#. The current "stage restriction" on splices using bindings from the current module is abolished.
+   Any stage n - 1 binding in a stage n splice is fair game.
+
+#. Relax ``-XTemplateHaskellQuotes`` to instead allow splices, but restrict their usage so all syntax is in stages >= 0.
+
+#. Introduce ``-XTemplateStagePersistence``.
+   Which is implied by ``-XTemplateHaskellQuotes`` (and thus plain ``-XTemplateHaskell``) for backwards compatibility.
+   It allows the current behavior where we blur the distinction between stages.
+   In particular, with `TemplateStagePersistence` enabled:
+
+   - Stage 0 identifiers bound in another module can be used in stage -1 (splices).
+   - Stage 0 identifiers bound at the top level can be used "by reference" in stage 1.
+   - Typed template haskell is allowed.
+   - The ``Lift`` type class and all its associated definitions are made available.
+   - Stage 0 identifiers bound anywhere can be used "by value" in stage 1, via an implicit ``lift``.
+
+   These are always permitted today.
+   But with ``-XNoTemplateStagePersistence``, overriding the default, all of those are *disabled*.
+
+#. Extend the command line [TODO bikeshed!!] with a way to specify per-stage package dependencies and the like.
+   If the emitted platform is specified without regards to a specific stage stage, it applies to stages 0, while stages -1 is left the same.
+   If the platform of stage other than those two isn't specified, it defaults to that of the stage next closest to 0.
+   [That's n takes's n + 1's, if n < -1, and n take's n - 1's, if n > 0.]
+   All that said, the emitted platform can still be specified per-stage like the other flags.
+
+#. When importing modules/packages, after applying the import offset ensure that the platforms match.
+   Note that while each imported module only has exports in its own stage 0, those exports can contain quotes of code in stages greater than 0.
+   Those stages > 0 (by the imported modules' numbering) need to also match.
+
+#. Just as GHC defines ``*_HOST_OS`` and similar CPP identifiers today, define ``*_BUILD_*`` ones if the current module has any stage -1 package imports, and ``*_TARGET_*`` if the current module has any stage 1 package imports.
+   Not always defining them helps people not use the wrong one, and improves the caching of builds (in principle at least).
+
+Cabal
+~~~~~~~~~~~~
+
+#. Extend the ``build-depends`` syntax with an optional stage integer offset parameter.
+   The default is stage 0.
+   N.B ``build-tool-depends`` can be thought of as a stage -1 executable dependencies list.
+   "Those executables are executed at build time, like top-evel splices, and so need to be built for the build platform."
+   `<https://github.com/haskell/cabal/issues/5411>`_ asks for a ``run-tool-depends`` which would be nothing but a stage 0 executable depends.
+   ``setup-depends`` can also be thought of as a stage -1 executable dependencies list.
```

Comment:
`setup-depends` is libraries needed to compile `Setup.hs`, not executables.

Also, this is perhaps overly pedantic, but IMO calling it `stage -1` is a bit misleading, as in principle the execution context of `Setup` can be arbitrarily decoupled from the execution context of `GHC` (e.g. one could imagine someone implementing the `Setup` interface backed by a heterogeneous-architecture build farm)

Reply:
```suggestion
   ``setup-depends`` can also be thought of as a stage -1 dependencies list.
   In both cases though, the metaphor is not perfect because separate executable are only loosely coupled.
   E.g libraries for Template-Haskell must be built with the same compiler / same ABI, but executable deps (including `./Setup` itself), need not be.
```

Reply:
Hows that?